# Unigrams v/s Bigrams v/s "UniBigrams?"

In [1]:
import pandas as pd

In [ ]:
data1 = pd.read_table("E:\\Yelp\\Unfiltered Data\\YelpZip\\reviewContent", header=None)
data2 = pd.read_table("E:\\Yelp\\Unfiltered Data\\YelpZip\\metadata", header=None)

data2.head()

In [ ]:
data = pd.merge(data1, data2, left_on=[0,2], right_on=[0,4], how='inner')

In [ ]:
data.head()
data.drop(2, inplace=True, axis=1)
data.drop('1_x', inplace=True, axis=1)
data.drop(4, inplace=True, axis=1)
data.head()

In [ ]:
data.head()
data.columns = ['id', 'date', 'text', 'item', 'rating', 'real']
data.head()

In [ ]:
data.to_csv("E:\\Yelp\\Unfiltered Data\\YelpZip\\Customs\\MAIN", header=None)

In [ ]:
data = data[(data['rating'] == 1) | (data['rating'] == 5)]

In [ ]:
real_1 = data[(data['rating'] == 1) & (data['real'] == 1)] 
real_5 = data[(data['rating'] == 5) & (data['real'] == 1)] 
fake_1 = data[(data['rating'] == 1) & (data['real'] == -1)] 
fake_5 = data[(data['rating'] == 5) & (data['real'] == -1)] 

real = real_1.append(real_5, ignore_index=True)
real.shape
fake = fake_1.append(fake_5, ignore_index=True)
fake.shape

In [ ]:
data = real.iloc[:10000,:].append(fake.iloc[:10000,:], ignore_index = True)

In [6]:
data.shape

(20000, 7)

In [2]:
data = pd.read_csv("E:\\Yelp\\Unfiltered Data\\YelpZip\\Customs\\SAMPLE", header=None)

In [10]:
from nltk.util import bigrams
import nltk

In [12]:
text = "Sample text. Just for fun!"
bigm = list(nltk.bigrams(text.split()))

In [13]:
bigm

[('Sample', 'text.'), ('text.', 'Just'), ('Just', 'for'), ('for', 'fun!')]

In [18]:
from nltk.corpus import stopwords
import string
stop_words=set(stopwords.words("english"))
print("Stopwords from the NLTK corpus: ", stop_words)

def text_process(text):
    #String punctuation provides all the necessary checks
    nopunc = [char for char in text if char not in string.punctuation]
    nopunc = ''.join(nopunc)
    return [word for word in nopunc.split() if word.lower() not in stopwords.words('english')]

import tqdm
tqdm.tqdm.pandas()
data[3] = data[3].progress_apply(lambda txt: text_process(txt))

Stopwords from the NLTK corpus:  {'doesn', 'before', "don't", 'because', 'against', 'about', 'below', "aren't", 'if', 'each', 'we', 'aren', 'the', 'for', 'be', 'yourselves', 'our', 'into', 'from', 'can', "hadn't", 'isn', 'with', "wouldn't", 'more', "should've", 'to', 'he', 'in', 'of', 'itself', 'd', 'theirs', 'too', 'she', 'all', "couldn't", 'not', 'y', 'a', 'myself', 'at', 'won', 'shouldn', 've', 'further', 'i', 'you', 'off', 'both', 'no', "shan't", 'its', 'there', 'are', 'why', 'wasn', 'mightn', 'will', 'during', 'and', 'them', "weren't", 'by', 'themselves', 'doing', 'after', 'they', 'having', "that'll", 'very', 'my', 'while', 'yourself', 's', "isn't", 'only', 'me', 'their', "hasn't", 'am', 'mustn', 'between', 'ours', 'when', 'couldn', 't', 'this', 'm', 'her', 're', "you've", 'was', 'up', 'didn', 'again', 'needn', 'which', 'his', 'through', 'that', 'were', 'don', 'above', 'than', 'now', 'until', 'being', "she's", 'does', 'own', 'once', 'him', 'your', "you'd", 'most', 'then', 'just', 



  0%|                                                                                        | 0/20000 [00:00<?, ?it/s]

  0%|                                                                                | 4/20000 [00:00<09:34, 34.79it/s]

  0%|                                                                                | 9/20000 [00:00<08:49, 37.76it/s]

  0%|                                                                               | 12/20000 [00:00<09:40, 34.42it/s]

  0%|                                                                               | 16/20000 [00:00<09:59, 33.35it/s]

  0%|                                                                               | 21/20000 [00:00<09:44, 34.19it/s]

  0%|                                                                               | 24/20000 [00:00<10:42, 31.11it/s]

  0%|                                                                               | 27/20000 [00:00<11:18, 29.44it/s]

  0%|                         

  2%|█▎                                                                            | 347/20000 [00:10<07:06, 46.13it/s]

  2%|█▍                                                                            | 354/20000 [00:10<06:57, 47.05it/s]

  2%|█▍                                                                            | 360/20000 [00:10<08:37, 37.98it/s]

  2%|█▍                                                                            | 367/20000 [00:10<07:31, 43.52it/s]

  2%|█▍                                                                            | 377/20000 [00:10<06:24, 51.02it/s]

  2%|█▍                                                                            | 384/20000 [00:11<08:07, 40.21it/s]

  2%|█▌                                                                            | 390/20000 [00:11<09:03, 36.08it/s]

  2%|█▌                                                                            | 395/20000 [00:11<08:34, 38.11it/s]

  2%|█▌                         

  4%|██▉                                                                           | 752/20000 [00:21<09:51, 32.55it/s]

  4%|██▉                                                                           | 756/20000 [00:21<11:46, 27.23it/s]

  4%|██▉                                                                           | 760/20000 [00:21<11:24, 28.13it/s]

  4%|██▉                                                                           | 764/20000 [00:21<12:33, 25.54it/s]

  4%|███                                                                           | 770/20000 [00:21<10:24, 30.80it/s]

  4%|███                                                                           | 774/20000 [00:22<09:41, 33.07it/s]

  4%|███                                                                           | 779/20000 [00:22<08:43, 36.72it/s]

  4%|███                                                                           | 786/20000 [00:22<07:50, 40.80it/s]

  4%|███                        

  5%|████▏                                                                        | 1094/20000 [00:31<09:22, 33.59it/s]

  5%|████▏                                                                        | 1098/20000 [00:31<11:10, 28.18it/s]

  6%|████▏                                                                        | 1102/20000 [00:32<12:38, 24.91it/s]

  6%|████▎                                                                        | 1106/20000 [00:32<11:33, 27.23it/s]

  6%|████▎                                                                        | 1113/20000 [00:32<09:33, 32.93it/s]

  6%|████▎                                                                        | 1118/20000 [00:32<10:13, 30.78it/s]

  6%|████▎                                                                        | 1122/20000 [00:32<10:03, 31.29it/s]

  6%|████▎                                                                        | 1127/20000 [00:32<09:24, 33.41it/s]

  6%|████▎                      

  7%|█████▌                                                                       | 1444/20000 [00:42<10:22, 29.83it/s]

  7%|█████▌                                                                       | 1448/20000 [00:42<10:14, 30.17it/s]

  7%|█████▌                                                                       | 1452/20000 [00:42<09:35, 32.25it/s]

  7%|█████▌                                                                       | 1456/20000 [00:42<09:44, 31.72it/s]

  7%|█████▌                                                                       | 1460/20000 [00:42<10:37, 29.09it/s]

  7%|█████▋                                                                       | 1464/20000 [00:42<09:50, 31.38it/s]

  7%|█████▋                                                                       | 1468/20000 [00:43<10:42, 28.83it/s]

  7%|█████▋                                                                       | 1473/20000 [00:43<09:36, 32.12it/s]

  7%|█████▋                     

  9%|███████▏                                                                     | 1882/20000 [00:55<05:39, 53.31it/s]

  9%|███████▎                                                                     | 1890/20000 [00:55<05:24, 55.86it/s]

  9%|███████▎                                                                     | 1897/20000 [00:55<05:19, 56.72it/s]

 10%|███████▎                                                                     | 1904/20000 [00:55<05:55, 50.94it/s]

 10%|███████▎                                                                     | 1910/20000 [00:55<05:44, 52.51it/s]

 10%|███████▍                                                                     | 1916/20000 [00:55<06:14, 48.34it/s]

 10%|███████▍                                                                     | 1922/20000 [00:56<09:03, 33.27it/s]

 10%|███████▍                                                                     | 1927/20000 [00:56<12:38, 23.82it/s]

 10%|███████▍                   

 11%|████████▌                                                                    | 2221/20000 [01:06<15:38, 18.94it/s]

 11%|████████▌                                                                    | 2224/20000 [01:07<14:07, 20.97it/s]

 11%|████████▌                                                                    | 2228/20000 [01:07<12:30, 23.68it/s]

 11%|████████▌                                                                    | 2231/20000 [01:07<11:50, 25.01it/s]

 11%|████████▌                                                                    | 2234/20000 [01:07<12:42, 23.30it/s]

 11%|████████▋                                                                    | 2245/20000 [01:07<09:42, 30.46it/s]

 11%|████████▋                                                                    | 2251/20000 [01:07<12:01, 24.58it/s]

 11%|████████▋                                                                    | 2256/20000 [01:08<11:50, 24.98it/s]

 11%|████████▋                  

 13%|█████████▋                                                                   | 2516/20000 [01:18<18:16, 15.95it/s]

 13%|█████████▋                                                                   | 2520/20000 [01:18<15:13, 19.14it/s]

 13%|█████████▋                                                                   | 2523/20000 [01:19<18:10, 16.02it/s]

 13%|█████████▋                                                                   | 2526/20000 [01:19<17:16, 16.86it/s]

 13%|█████████▋                                                                   | 2530/20000 [01:19<14:28, 20.12it/s]

 13%|█████████▊                                                                   | 2533/20000 [01:19<14:55, 19.51it/s]

 13%|█████████▊                                                                   | 2536/20000 [01:19<15:23, 18.90it/s]

 13%|█████████▊                                                                   | 2539/20000 [01:19<14:13, 20.47it/s]

 13%|█████████▊                 

 14%|██████████▋                                                                  | 2790/20000 [01:29<10:56, 26.20it/s]

 14%|██████████▊                                                                  | 2793/20000 [01:29<11:48, 24.30it/s]

 14%|██████████▊                                                                  | 2796/20000 [01:29<12:44, 22.50it/s]

 14%|██████████▊                                                                  | 2799/20000 [01:30<12:47, 22.41it/s]

 14%|██████████▊                                                                  | 2803/20000 [01:30<13:38, 21.01it/s]

 14%|██████████▊                                                                  | 2806/20000 [01:30<16:02, 17.87it/s]

 14%|██████████▊                                                                  | 2808/20000 [01:30<20:31, 13.96it/s]

 14%|██████████▊                                                                  | 2813/20000 [01:30<16:15, 17.62it/s]

 14%|██████████▊                

 15%|███████████▊                                                                 | 3082/20000 [01:40<14:20, 19.65it/s]

 15%|███████████▉                                                                 | 3086/20000 [01:40<12:15, 22.99it/s]

 15%|███████████▉                                                                 | 3089/20000 [01:40<11:52, 23.72it/s]

 15%|███████████▉                                                                 | 3093/20000 [01:40<10:55, 25.80it/s]

 15%|███████████▉                                                                 | 3096/20000 [01:40<12:25, 22.68it/s]

 16%|███████████▉                                                                 | 3103/20000 [01:40<09:56, 28.34it/s]

 16%|███████████▉                                                                 | 3108/20000 [01:41<08:56, 31.50it/s]

 16%|███████████▉                                                                 | 3112/20000 [01:41<12:04, 23.33it/s]

 16%|███████████▉               

 17%|█████████████▏                                                               | 3427/20000 [01:52<05:51, 47.19it/s]

 17%|█████████████▏                                                               | 3434/20000 [01:52<06:43, 41.02it/s]

 17%|█████████████▏                                                               | 3440/20000 [01:52<06:50, 40.32it/s]

 17%|█████████████▎                                                               | 3446/20000 [01:53<11:12, 24.62it/s]

 17%|█████████████▎                                                               | 3451/20000 [01:53<12:27, 22.13it/s]

 17%|█████████████▎                                                               | 3463/20000 [01:53<09:28, 29.08it/s]

 17%|█████████████▎                                                               | 3469/20000 [01:53<11:31, 23.90it/s]

 17%|█████████████▎                                                               | 3474/20000 [01:54<13:34, 20.28it/s]

 17%|█████████████▍             

 19%|██████████████▌                                                              | 3784/20000 [02:03<07:07, 37.96it/s]

 19%|██████████████▌                                                              | 3789/20000 [02:03<07:23, 36.54it/s]

 19%|██████████████▌                                                              | 3794/20000 [02:03<07:02, 38.31it/s]

 19%|██████████████▋                                                              | 3806/20000 [02:04<05:57, 45.27it/s]

 19%|██████████████▋                                                              | 3812/20000 [02:04<06:41, 40.28it/s]

 19%|██████████████▋                                                              | 3817/20000 [02:04<11:51, 22.73it/s]

 19%|██████████████▋                                                              | 3821/20000 [02:04<11:29, 23.46it/s]

 19%|██████████████▋                                                              | 3827/20000 [02:04<09:44, 27.66it/s]

 19%|██████████████▊            

 21%|███████████████▉                                                             | 4140/20000 [02:14<10:07, 26.12it/s]

 21%|███████████████▉                                                             | 4147/20000 [02:14<08:30, 31.04it/s]

 21%|███████████████▉                                                             | 4152/20000 [02:14<08:09, 32.35it/s]

 21%|████████████████                                                             | 4157/20000 [02:14<08:38, 30.53it/s]

 21%|████████████████                                                             | 4161/20000 [02:14<09:29, 27.82it/s]

 21%|████████████████                                                             | 4167/20000 [02:14<08:04, 32.70it/s]

 21%|████████████████                                                             | 4172/20000 [02:15<08:13, 32.10it/s]

 21%|████████████████                                                             | 4176/20000 [02:15<09:10, 28.76it/s]

 21%|████████████████           

 22%|█████████████████                                                            | 4430/20000 [02:25<15:46, 16.46it/s]

 22%|█████████████████                                                            | 4433/20000 [02:25<14:16, 18.17it/s]

 22%|█████████████████                                                            | 4436/20000 [02:25<13:22, 19.40it/s]

 22%|█████████████████                                                            | 4439/20000 [02:26<13:07, 19.77it/s]

 22%|█████████████████                                                            | 4442/20000 [02:26<13:16, 19.53it/s]

 22%|█████████████████                                                            | 4445/20000 [02:26<12:42, 20.39it/s]

 22%|█████████████████                                                            | 4448/20000 [02:26<14:00, 18.50it/s]

 22%|█████████████████▏                                                           | 4450/20000 [02:26<15:47, 16.41it/s]

 22%|█████████████████▏         

 23%|██████████████████                                                           | 4697/20000 [02:36<11:24, 22.36it/s]

 24%|██████████████████                                                           | 4701/20000 [02:36<10:30, 24.27it/s]

 24%|██████████████████                                                           | 4704/20000 [02:36<11:44, 21.70it/s]

 24%|██████████████████                                                           | 4707/20000 [02:36<11:26, 22.27it/s]

 24%|██████████████████▏                                                          | 4712/20000 [02:36<09:48, 25.96it/s]

 24%|██████████████████▏                                                          | 4717/20000 [02:36<08:42, 29.26it/s]

 24%|██████████████████▏                                                          | 4721/20000 [02:37<08:47, 28.99it/s]

 24%|██████████████████▏                                                          | 4728/20000 [02:37<07:49, 32.50it/s]

 24%|██████████████████▏        

 25%|███████████████████▎                                                         | 5027/20000 [02:46<09:42, 25.69it/s]

 25%|███████████████████▎                                                         | 5030/20000 [02:46<10:01, 24.90it/s]

 25%|███████████████████▍                                                         | 5033/20000 [02:46<09:48, 25.42it/s]

 25%|███████████████████▍                                                         | 5036/20000 [02:46<11:04, 22.50it/s]

 25%|███████████████████▍                                                         | 5040/20000 [02:47<11:31, 21.65it/s]

 25%|███████████████████▍                                                         | 5045/20000 [02:47<10:10, 24.51it/s]

 25%|███████████████████▍                                                         | 5048/20000 [02:47<09:51, 25.28it/s]

 25%|███████████████████▍                                                         | 5053/20000 [02:47<08:40, 28.70it/s]

 25%|███████████████████▍       

 27%|████████████████████▍                                                        | 5319/20000 [02:56<06:18, 38.76it/s]

 27%|████████████████████▍                                                        | 5324/20000 [02:56<06:39, 36.71it/s]

 27%|████████████████████▌                                                        | 5329/20000 [02:56<07:47, 31.41it/s]

 27%|████████████████████▌                                                        | 5333/20000 [02:56<08:28, 28.86it/s]

 27%|████████████████████▌                                                        | 5339/20000 [02:57<07:20, 33.27it/s]

 27%|████████████████████▌                                                        | 5344/20000 [02:57<07:15, 33.64it/s]

 27%|████████████████████▌                                                        | 5349/20000 [02:57<07:14, 33.68it/s]

 27%|████████████████████▌                                                        | 5353/20000 [02:57<07:31, 32.44it/s]

 27%|████████████████████▋      

 28%|█████████████████████▊                                                       | 5677/20000 [03:06<12:36, 18.93it/s]

 28%|█████████████████████▉                                                       | 5683/20000 [03:06<10:20, 23.08it/s]

 28%|█████████████████████▉                                                       | 5688/20000 [03:06<10:19, 23.10it/s]

 28%|█████████████████████▉                                                       | 5693/20000 [03:07<09:15, 25.76it/s]

 28%|█████████████████████▉                                                       | 5697/20000 [03:07<11:25, 20.86it/s]

 28%|█████████████████████▉                                                       | 5700/20000 [03:07<12:13, 19.50it/s]

 29%|█████████████████████▉                                                       | 5704/20000 [03:07<10:39, 22.34it/s]

 29%|█████████████████████▉                                                       | 5709/20000 [03:07<09:10, 25.95it/s]

 29%|█████████████████████▉     

 30%|██████████████████████▉                                                      | 5962/20000 [03:18<06:56, 33.70it/s]

 30%|██████████████████████▉                                                      | 5967/20000 [03:18<06:48, 34.37it/s]

 30%|██████████████████████▉                                                      | 5972/20000 [03:18<08:09, 28.67it/s]

 30%|███████████████████████                                                      | 5976/20000 [03:18<08:22, 27.92it/s]

 30%|███████████████████████                                                      | 5980/20000 [03:19<11:25, 20.44it/s]

 30%|███████████████████████                                                      | 5983/20000 [03:19<10:39, 21.91it/s]

 30%|███████████████████████                                                      | 5992/20000 [03:19<08:56, 26.11it/s]

 30%|███████████████████████                                                      | 5998/20000 [03:19<07:55, 29.45it/s]

 30%|███████████████████████    

 31%|████████████████████████▏                                                    | 6291/20000 [03:30<14:41, 15.55it/s]

 31%|████████████████████████▏                                                    | 6294/20000 [03:30<14:26, 15.82it/s]

 31%|████████████████████████▏                                                    | 6297/20000 [03:30<12:50, 17.78it/s]

 32%|████████████████████████▎                                                    | 6300/20000 [03:30<14:42, 15.53it/s]

 32%|████████████████████████▎                                                    | 6305/20000 [03:30<11:45, 19.42it/s]

 32%|████████████████████████▎                                                    | 6308/20000 [03:31<14:04, 16.21it/s]

 32%|████████████████████████▎                                                    | 6311/20000 [03:31<16:31, 13.81it/s]

 32%|████████████████████████▎                                                    | 6315/20000 [03:31<13:30, 16.89it/s]

 32%|████████████████████████▎  

 33%|█████████████████████████▎                                                   | 6584/20000 [03:40<08:35, 26.04it/s]

 33%|█████████████████████████▎                                                   | 6587/20000 [03:41<10:47, 20.73it/s]

 33%|█████████████████████████▍                                                   | 6593/20000 [03:41<08:40, 25.73it/s]

 33%|█████████████████████████▍                                                   | 6597/20000 [03:41<09:31, 23.44it/s]

 33%|█████████████████████████▍                                                   | 6601/20000 [03:41<09:04, 24.62it/s]

 33%|█████████████████████████▍                                                   | 6606/20000 [03:41<07:53, 28.29it/s]

 33%|█████████████████████████▍                                                   | 6610/20000 [03:41<07:30, 29.70it/s]

 33%|█████████████████████████▍                                                   | 6614/20000 [03:42<08:56, 24.97it/s]

 33%|█████████████████████████▍ 

 34%|██████████████████████████▍                                                  | 6870/20000 [03:51<05:09, 42.39it/s]

 34%|██████████████████████████▍                                                  | 6875/20000 [03:51<04:58, 43.98it/s]

 34%|██████████████████████████▍                                                  | 6882/20000 [03:51<04:26, 49.17it/s]

 34%|██████████████████████████▌                                                  | 6889/20000 [03:51<04:13, 51.81it/s]

 34%|██████████████████████████▌                                                  | 6895/20000 [03:51<05:18, 41.14it/s]

 34%|██████████████████████████▌                                                  | 6900/20000 [03:52<05:48, 37.57it/s]

 35%|██████████████████████████▌                                                  | 6905/20000 [03:52<06:26, 33.90it/s]

 35%|██████████████████████████▌                                                  | 6909/20000 [03:52<07:57, 27.42it/s]

 35%|██████████████████████████▌

 36%|███████████████████████████▌                                                 | 7171/20000 [04:01<08:24, 25.42it/s]

 36%|███████████████████████████▋                                                 | 7177/20000 [04:01<07:15, 29.41it/s]

 36%|███████████████████████████▋                                                 | 7181/20000 [04:02<08:18, 25.71it/s]

 36%|███████████████████████████▋                                                 | 7185/20000 [04:02<09:07, 23.43it/s]

 36%|███████████████████████████▋                                                 | 7192/20000 [04:02<07:49, 27.28it/s]

 36%|███████████████████████████▋                                                 | 7196/20000 [04:02<07:40, 27.81it/s]

 36%|███████████████████████████▋                                                 | 7201/20000 [04:02<06:48, 31.30it/s]

 36%|███████████████████████████▋                                                 | 7205/20000 [04:02<07:30, 28.39it/s]

 36%|███████████████████████████

 37%|████████████████████████████▊                                                | 7496/20000 [04:11<06:49, 30.50it/s]

 38%|████████████████████████████▉                                                | 7500/20000 [04:12<10:41, 19.49it/s]

 38%|████████████████████████████▉                                                | 7503/20000 [04:12<09:51, 21.13it/s]

 38%|████████████████████████████▉                                                | 7507/20000 [04:12<08:49, 23.61it/s]

 38%|████████████████████████████▉                                                | 7512/20000 [04:12<07:59, 26.06it/s]

 38%|████████████████████████████▉                                                | 7516/20000 [04:12<09:28, 21.96it/s]

 38%|████████████████████████████▉                                                | 7520/20000 [04:12<09:11, 22.63it/s]

 38%|████████████████████████████▉                                                | 7523/20000 [04:13<08:31, 24.38it/s]

 38%|███████████████████████████

 39%|██████████████████████████████                                               | 7802/20000 [04:22<05:43, 35.56it/s]

 39%|██████████████████████████████                                               | 7806/20000 [04:22<06:59, 29.06it/s]

 39%|██████████████████████████████                                               | 7810/20000 [04:22<08:02, 25.27it/s]

 39%|██████████████████████████████                                               | 7814/20000 [04:22<07:31, 27.01it/s]

 39%|██████████████████████████████                                               | 7818/20000 [04:22<07:08, 28.44it/s]

 39%|██████████████████████████████                                               | 7822/20000 [04:22<07:43, 26.28it/s]

 39%|██████████████████████████████▏                                              | 7826/20000 [04:23<07:31, 26.96it/s]

 39%|██████████████████████████████▏                                              | 7830/20000 [04:23<06:58, 29.10it/s]

 39%|███████████████████████████

 41%|███████████████████████████████▎                                             | 8118/20000 [04:33<06:55, 28.63it/s]

 41%|███████████████████████████████▎                                             | 8123/20000 [04:33<06:40, 29.69it/s]

 41%|███████████████████████████████▎                                             | 8129/20000 [04:33<06:03, 32.64it/s]

 41%|███████████████████████████████▎                                             | 8133/20000 [04:33<06:33, 30.16it/s]

 41%|███████████████████████████████▎                                             | 8137/20000 [04:33<06:42, 29.46it/s]

 41%|███████████████████████████████▎                                             | 8143/20000 [04:33<05:52, 33.66it/s]

 41%|███████████████████████████████▎                                             | 8147/20000 [04:34<06:15, 31.57it/s]

 41%|███████████████████████████████▍                                             | 8151/20000 [04:34<06:30, 30.31it/s]

 41%|███████████████████████████

 42%|████████████████████████████████▍                                            | 8431/20000 [04:42<04:28, 43.04it/s]

 42%|████████████████████████████████▍                                            | 8439/20000 [04:43<04:17, 44.83it/s]

 42%|████████████████████████████████▌                                            | 8444/20000 [04:43<04:55, 39.17it/s]

 42%|████████████████████████████████▌                                            | 8449/20000 [04:43<05:31, 34.79it/s]

 42%|████████████████████████████████▌                                            | 8453/20000 [04:43<07:27, 25.78it/s]

 42%|████████████████████████████████▌                                            | 8457/20000 [04:43<06:54, 27.83it/s]

 42%|████████████████████████████████▌                                            | 8464/20000 [04:43<06:06, 31.48it/s]

 42%|████████████████████████████████▌                                            | 8468/20000 [04:44<06:07, 31.41it/s]

 42%|███████████████████████████

 44%|█████████████████████████████████▌                                           | 8709/20000 [04:53<10:16, 18.31it/s]

 44%|█████████████████████████████████▌                                           | 8712/20000 [04:53<11:18, 16.63it/s]

 44%|█████████████████████████████████▌                                           | 8715/20000 [04:53<10:08, 18.55it/s]

 44%|█████████████████████████████████▌                                           | 8719/20000 [04:54<08:52, 21.19it/s]

 44%|█████████████████████████████████▌                                           | 8722/20000 [04:54<08:30, 22.10it/s]

 44%|█████████████████████████████████▌                                           | 8725/20000 [04:54<09:12, 20.41it/s]

 44%|█████████████████████████████████▌                                           | 8733/20000 [04:54<07:29, 25.09it/s]

 44%|█████████████████████████████████▋                                           | 8737/20000 [04:54<07:43, 24.32it/s]

 44%|███████████████████████████

 45%|██████████████████████████████████▋                                          | 9010/20000 [05:04<05:41, 32.13it/s]

 45%|██████████████████████████████████▋                                          | 9017/20000 [05:04<04:58, 36.84it/s]

 45%|██████████████████████████████████▋                                          | 9022/20000 [05:04<05:17, 34.59it/s]

 45%|██████████████████████████████████▊                                          | 9026/20000 [05:04<05:59, 30.52it/s]

 45%|██████████████████████████████████▊                                          | 9030/20000 [05:04<05:59, 30.53it/s]

 45%|██████████████████████████████████▊                                          | 9040/20000 [05:05<04:45, 38.43it/s]

 45%|██████████████████████████████████▊                                          | 9046/20000 [05:05<05:40, 32.20it/s]

 45%|██████████████████████████████████▊                                          | 9051/20000 [05:05<06:41, 27.25it/s]

 45%|███████████████████████████

 46%|███████████████████████████████████▊                                         | 9298/20000 [05:15<13:53, 12.84it/s]

 47%|███████████████████████████████████▊                                         | 9301/20000 [05:15<12:27, 14.32it/s]

 47%|███████████████████████████████████▊                                         | 9306/20000 [05:15<09:49, 18.14it/s]

 47%|███████████████████████████████████▊                                         | 9309/20000 [05:16<08:57, 19.87it/s]

 47%|███████████████████████████████████▊                                         | 9314/20000 [05:16<07:27, 23.91it/s]

 47%|███████████████████████████████████▉                                         | 9320/20000 [05:16<06:10, 28.83it/s]

 47%|███████████████████████████████████▉                                         | 9325/20000 [05:16<05:30, 32.25it/s]

 47%|███████████████████████████████████▉                                         | 9330/20000 [05:16<05:41, 31.22it/s]

 47%|███████████████████████████

 48%|████████████████████████████████████▉                                        | 9585/20000 [05:26<07:05, 24.48it/s]

 48%|████████████████████████████████████▉                                        | 9588/20000 [05:27<11:52, 14.60it/s]

 48%|████████████████████████████████████▉                                        | 9594/20000 [05:27<09:31, 18.21it/s]

 48%|████████████████████████████████████▉                                        | 9597/20000 [05:27<10:02, 17.26it/s]

 48%|████████████████████████████████████▉                                        | 9600/20000 [05:27<09:20, 18.56it/s]

 48%|████████████████████████████████████▉                                        | 9603/20000 [05:27<08:21, 20.74it/s]

 48%|████████████████████████████████████▉                                        | 9608/20000 [05:27<07:22, 23.51it/s]

 48%|█████████████████████████████████████                                        | 9611/20000 [05:28<07:36, 22.78it/s]

 48%|███████████████████████████

 49%|█████████████████████████████████████▉                                       | 9850/20000 [05:38<08:24, 20.14it/s]

 49%|█████████████████████████████████████▉                                       | 9856/20000 [05:38<06:45, 25.04it/s]

 49%|█████████████████████████████████████▉                                       | 9860/20000 [05:38<07:53, 21.43it/s]

 49%|█████████████████████████████████████▉                                       | 9863/20000 [05:38<07:50, 21.56it/s]

 49%|█████████████████████████████████████▉                                       | 9867/20000 [05:38<06:48, 24.78it/s]

 49%|██████████████████████████████████████                                       | 9873/20000 [05:39<06:02, 27.97it/s]

 49%|██████████████████████████████████████                                       | 9878/20000 [05:39<05:51, 28.77it/s]

 49%|██████████████████████████████████████                                       | 9882/20000 [05:39<08:09, 20.69it/s]

 49%|███████████████████████████

 51%|██████████████████████████████████████▉                                     | 10262/20000 [05:49<04:30, 35.96it/s]

 51%|███████████████████████████████████████                                     | 10267/20000 [05:49<04:20, 37.32it/s]

 51%|███████████████████████████████████████                                     | 10275/20000 [05:49<03:41, 44.00it/s]

 51%|███████████████████████████████████████                                     | 10281/20000 [05:49<03:26, 47.04it/s]

 51%|███████████████████████████████████████                                     | 10287/20000 [05:50<05:15, 30.81it/s]

 51%|███████████████████████████████████████                                     | 10292/20000 [05:50<04:42, 34.31it/s]

 51%|███████████████████████████████████████▏                                    | 10297/20000 [05:50<05:04, 31.85it/s]

 52%|███████████████████████████████████████▏                                    | 10301/20000 [05:50<04:52, 33.15it/s]

 52%|███████████████████████████

 53%|████████████████████████████████████████▏                                   | 10587/20000 [06:01<07:55, 19.81it/s]

 53%|████████████████████████████████████████▏                                   | 10590/20000 [06:01<07:40, 20.43it/s]

 53%|████████████████████████████████████████▎                                   | 10593/20000 [06:01<07:56, 19.75it/s]

 53%|████████████████████████████████████████▎                                   | 10596/20000 [06:01<07:12, 21.73it/s]

 53%|████████████████████████████████████████▎                                   | 10599/20000 [06:01<07:19, 21.39it/s]

 53%|████████████████████████████████████████▎                                   | 10602/20000 [06:01<07:47, 20.09it/s]

 53%|████████████████████████████████████████▎                                   | 10605/20000 [06:02<08:19, 18.81it/s]

 53%|████████████████████████████████████████▎                                   | 10607/20000 [06:02<08:34, 18.26it/s]

 53%|███████████████████████████

 55%|█████████████████████████████████████████▋                                  | 10960/20000 [06:12<04:07, 36.56it/s]

 55%|█████████████████████████████████████████▋                                  | 10965/20000 [06:12<06:24, 23.48it/s]

 55%|█████████████████████████████████████████▋                                  | 10970/20000 [06:12<05:25, 27.74it/s]

 55%|█████████████████████████████████████████▋                                  | 10974/20000 [06:12<06:06, 24.62it/s]

 55%|█████████████████████████████████████████▋                                  | 10978/20000 [06:12<06:08, 24.45it/s]

 55%|█████████████████████████████████████████▋                                  | 10981/20000 [06:13<06:36, 22.75it/s]

 55%|█████████████████████████████████████████▋                                  | 10984/20000 [06:13<06:10, 24.33it/s]

 55%|█████████████████████████████████████████▊                                  | 10988/20000 [06:13<05:35, 26.84it/s]

 55%|███████████████████████████

 56%|██████████████████████████████████████████▍                                 | 11176/20000 [06:23<10:56, 13.45it/s]

 56%|██████████████████████████████████████████▍                                 | 11179/20000 [06:23<09:11, 16.01it/s]

 56%|██████████████████████████████████████████▍                                 | 11182/20000 [06:23<08:21, 17.57it/s]

 56%|██████████████████████████████████████████▌                                 | 11185/20000 [06:24<08:25, 17.45it/s]

 56%|██████████████████████████████████████████▌                                 | 11187/20000 [06:24<10:10, 14.43it/s]

 56%|██████████████████████████████████████████▌                                 | 11189/20000 [06:24<13:20, 11.01it/s]

 56%|██████████████████████████████████████████▌                                 | 11191/20000 [06:24<14:28, 10.15it/s]

 56%|██████████████████████████████████████████▌                                 | 11193/20000 [06:24<12:56, 11.33it/s]

 56%|███████████████████████████

 57%|███████████████████████████████████████████▋                                | 11485/20000 [06:34<03:29, 40.70it/s]

 57%|███████████████████████████████████████████▋                                | 11494/20000 [06:34<02:56, 48.14it/s]

 57%|███████████████████████████████████████████▋                                | 11500/20000 [06:34<04:02, 35.04it/s]

 58%|███████████████████████████████████████████▋                                | 11505/20000 [06:34<03:53, 36.45it/s]

 58%|███████████████████████████████████████████▋                                | 11510/20000 [06:34<04:38, 30.50it/s]

 58%|███████████████████████████████████████████▊                                | 11516/20000 [06:35<03:57, 35.77it/s]

 58%|███████████████████████████████████████████▊                                | 11521/20000 [06:35<04:43, 29.89it/s]

 58%|███████████████████████████████████████████▊                                | 11525/20000 [06:35<05:03, 27.95it/s]

 58%|███████████████████████████

 59%|████████████████████████████████████████████▉                               | 11834/20000 [06:44<04:22, 31.05it/s]

 59%|████████████████████████████████████████████▉                               | 11838/20000 [06:44<04:23, 30.96it/s]

 59%|████████████████████████████████████████████▉                               | 11842/20000 [06:45<05:30, 24.66it/s]

 59%|█████████████████████████████████████████████                               | 11846/20000 [06:45<05:03, 26.90it/s]

 59%|█████████████████████████████████████████████                               | 11850/20000 [06:45<04:48, 28.29it/s]

 59%|█████████████████████████████████████████████                               | 11854/20000 [06:45<04:54, 27.66it/s]

 59%|█████████████████████████████████████████████                               | 11857/20000 [06:45<05:16, 25.74it/s]

 59%|█████████████████████████████████████████████                               | 11863/20000 [06:45<04:30, 30.04it/s]

 59%|███████████████████████████

 61%|██████████████████████████████████████████████▏                             | 12171/20000 [06:55<05:13, 24.99it/s]

 61%|██████████████████████████████████████████████▎                             | 12175/20000 [06:55<06:24, 20.36it/s]

 61%|██████████████████████████████████████████████▎                             | 12179/20000 [06:55<05:48, 22.44it/s]

 61%|██████████████████████████████████████████████▎                             | 12182/20000 [06:55<05:45, 22.61it/s]

 61%|██████████████████████████████████████████████▎                             | 12185/20000 [06:55<05:36, 23.21it/s]

 61%|██████████████████████████████████████████████▎                             | 12192/20000 [06:55<04:31, 28.81it/s]

 61%|██████████████████████████████████████████████▎                             | 12197/20000 [06:55<03:58, 32.78it/s]

 61%|██████████████████████████████████████████████▍                             | 12204/20000 [06:56<03:25, 37.92it/s]

 61%|███████████████████████████

 63%|███████████████████████████████████████████████▌                            | 12530/20000 [07:05<04:54, 25.35it/s]

 63%|███████████████████████████████████████████████▋                            | 12536/20000 [07:06<04:19, 28.72it/s]

 63%|███████████████████████████████████████████████▋                            | 12541/20000 [07:06<04:57, 25.08it/s]

 63%|███████████████████████████████████████████████▋                            | 12545/20000 [07:06<04:45, 26.07it/s]

 63%|███████████████████████████████████████████████▋                            | 12549/20000 [07:06<04:53, 25.38it/s]

 63%|███████████████████████████████████████████████▊                            | 12567/20000 [07:06<03:39, 33.90it/s]

 63%|███████████████████████████████████████████████▊                            | 12574/20000 [07:06<03:28, 35.68it/s]

 63%|███████████████████████████████████████████████▊                            | 12581/20000 [07:07<03:04, 40.17it/s]

 63%|███████████████████████████

 64%|████████████████████████████████████████████████▉                           | 12891/20000 [07:16<03:03, 38.71it/s]

 64%|█████████████████████████████████████████████████                           | 12897/20000 [07:16<02:48, 42.21it/s]

 65%|█████████████████████████████████████████████████                           | 12902/20000 [07:16<02:45, 42.91it/s]

 65%|█████████████████████████████████████████████████                           | 12907/20000 [07:16<03:00, 39.19it/s]

 65%|█████████████████████████████████████████████████                           | 12912/20000 [07:16<03:24, 34.73it/s]

 65%|█████████████████████████████████████████████████                           | 12916/20000 [07:16<03:33, 33.19it/s]

 65%|█████████████████████████████████████████████████                           | 12920/20000 [07:16<03:51, 30.61it/s]

 65%|█████████████████████████████████████████████████                           | 12924/20000 [07:17<04:14, 27.82it/s]

 65%|███████████████████████████

 66%|██████████████████████████████████████████████████▏                         | 13222/20000 [07:26<03:15, 34.61it/s]

 66%|██████████████████████████████████████████████████▎                         | 13226/20000 [07:27<03:19, 34.02it/s]

 66%|██████████████████████████████████████████████████▎                         | 13230/20000 [07:27<03:46, 29.86it/s]

 66%|██████████████████████████████████████████████████▎                         | 13234/20000 [07:27<03:56, 28.57it/s]

 66%|██████████████████████████████████████████████████▎                         | 13240/20000 [07:27<03:22, 33.43it/s]

 66%|██████████████████████████████████████████████████▎                         | 13245/20000 [07:27<03:03, 36.86it/s]

 66%|██████████████████████████████████████████████████▎                         | 13251/20000 [07:27<02:41, 41.68it/s]

 66%|██████████████████████████████████████████████████▎                         | 13256/20000 [07:27<03:33, 31.56it/s]

 66%|███████████████████████████

 68%|███████████████████████████████████████████████████▍                        | 13526/20000 [07:37<04:59, 21.61it/s]

 68%|███████████████████████████████████████████████████▍                        | 13529/20000 [07:37<05:32, 19.49it/s]

 68%|███████████████████████████████████████████████████▍                        | 13532/20000 [07:37<05:58, 18.05it/s]

 68%|███████████████████████████████████████████████████▍                        | 13535/20000 [07:37<06:02, 17.84it/s]

 68%|███████████████████████████████████████████████████▍                        | 13538/20000 [07:38<06:15, 17.19it/s]

 68%|███████████████████████████████████████████████████▍                        | 13540/20000 [07:38<06:29, 16.57it/s]

 68%|███████████████████████████████████████████████████▍                        | 13545/20000 [07:38<05:21, 20.08it/s]

 68%|███████████████████████████████████████████████████▍                        | 13548/20000 [07:38<04:52, 22.03it/s]

 68%|███████████████████████████

 69%|████████████████████████████████████████████████████▌                       | 13834/20000 [07:47<03:20, 30.73it/s]

 69%|████████████████████████████████████████████████████▌                       | 13838/20000 [07:47<03:35, 28.62it/s]

 69%|████████████████████████████████████████████████████▌                       | 13842/20000 [07:48<04:03, 25.29it/s]

 69%|████████████████████████████████████████████████████▌                       | 13845/20000 [07:48<03:58, 25.83it/s]

 69%|████████████████████████████████████████████████████▌                       | 13848/20000 [07:48<03:51, 26.59it/s]

 69%|████████████████████████████████████████████████████▋                       | 13851/20000 [07:48<04:24, 23.26it/s]

 69%|████████████████████████████████████████████████████▋                       | 13854/20000 [07:48<05:19, 19.21it/s]

 69%|████████████████████████████████████████████████████▋                       | 13858/20000 [07:49<05:48, 17.61it/s]

 69%|███████████████████████████

 71%|█████████████████████████████████████████████████████▌                      | 14101/20000 [07:59<03:37, 27.15it/s]

 71%|█████████████████████████████████████████████████████▌                      | 14106/20000 [07:59<03:31, 27.82it/s]

 71%|█████████████████████████████████████████████████████▌                      | 14110/20000 [07:59<04:04, 24.07it/s]

 71%|█████████████████████████████████████████████████████▋                      | 14113/20000 [07:59<06:08, 15.98it/s]

 71%|█████████████████████████████████████████████████████▋                      | 14116/20000 [07:59<06:03, 16.19it/s]

 71%|█████████████████████████████████████████████████████▋                      | 14119/20000 [08:00<07:06, 13.79it/s]

 71%|█████████████████████████████████████████████████████▋                      | 14123/20000 [08:00<05:58, 16.40it/s]

 71%|█████████████████████████████████████████████████████▋                      | 14126/20000 [08:00<05:34, 17.55it/s]

 71%|███████████████████████████

 72%|██████████████████████████████████████████████████████▌                     | 14355/20000 [08:09<03:07, 30.09it/s]

 72%|██████████████████████████████████████████████████████▌                     | 14359/20000 [08:09<03:17, 28.57it/s]

 72%|██████████████████████████████████████████████████████▌                     | 14363/20000 [08:09<03:09, 29.72it/s]

 72%|██████████████████████████████████████████████████████▌                     | 14367/20000 [08:10<03:29, 26.93it/s]

 72%|██████████████████████████████████████████████████████▌                     | 14372/20000 [08:10<03:00, 31.13it/s]

 72%|██████████████████████████████████████████████████████▋                     | 14377/20000 [08:10<02:53, 32.42it/s]

 72%|██████████████████████████████████████████████████████▋                     | 14382/20000 [08:10<02:48, 33.25it/s]

 72%|██████████████████████████████████████████████████████▋                     | 14386/20000 [08:10<05:05, 18.40it/s]

 72%|███████████████████████████

 73%|███████████████████████████████████████████████████████▋                    | 14644/20000 [08:20<03:43, 23.95it/s]

 73%|███████████████████████████████████████████████████████▋                    | 14647/20000 [08:20<03:40, 24.25it/s]

 73%|███████████████████████████████████████████████████████▋                    | 14657/20000 [08:20<02:51, 31.23it/s]

 73%|███████████████████████████████████████████████████████▋                    | 14667/20000 [08:20<02:16, 39.20it/s]

 73%|███████████████████████████████████████████████████████▊                    | 14677/20000 [08:21<01:53, 47.05it/s]

 73%|███████████████████████████████████████████████████████▊                    | 14685/20000 [08:21<01:47, 49.47it/s]

 73%|███████████████████████████████████████████████████████▊                    | 14692/20000 [08:21<01:57, 45.22it/s]

 73%|███████████████████████████████████████████████████████▊                    | 14698/20000 [08:21<02:29, 35.53it/s]

 74%|███████████████████████████

 75%|████████████████████████████████████████████████████████▉                   | 14990/20000 [08:31<02:35, 32.14it/s]

 75%|████████████████████████████████████████████████████████▉                   | 14994/20000 [08:31<02:44, 30.51it/s]

 75%|████████████████████████████████████████████████████████▉                   | 14998/20000 [08:31<02:59, 27.89it/s]

 75%|█████████████████████████████████████████████████████████                   | 15004/20000 [08:31<02:32, 32.65it/s]

 75%|█████████████████████████████████████████████████████████                   | 15008/20000 [08:31<02:40, 31.09it/s]

 75%|█████████████████████████████████████████████████████████                   | 15012/20000 [08:31<03:01, 27.46it/s]

 75%|█████████████████████████████████████████████████████████                   | 15016/20000 [08:32<03:18, 25.10it/s]

 75%|█████████████████████████████████████████████████████████                   | 15020/20000 [08:32<03:02, 27.33it/s]

 75%|███████████████████████████

 76%|██████████████████████████████████████████████████████████                  | 15281/20000 [08:41<02:16, 34.58it/s]

 76%|██████████████████████████████████████████████████████████                  | 15286/20000 [08:42<03:19, 23.68it/s]

 76%|██████████████████████████████████████████████████████████                  | 15290/20000 [08:42<03:06, 25.24it/s]

 76%|██████████████████████████████████████████████████████████                  | 15294/20000 [08:42<03:08, 25.02it/s]

 76%|██████████████████████████████████████████████████████████▏                 | 15298/20000 [08:42<02:56, 26.66it/s]

 77%|██████████████████████████████████████████████████████████▏                 | 15303/20000 [08:42<02:48, 27.88it/s]

 77%|██████████████████████████████████████████████████████████▏                 | 15307/20000 [08:42<03:03, 25.54it/s]

 77%|██████████████████████████████████████████████████████████▏                 | 15311/20000 [08:42<02:47, 28.07it/s]

 77%|███████████████████████████

 78%|███████████████████████████████████████████████████████████▏                | 15574/20000 [08:52<03:17, 22.37it/s]

 78%|███████████████████████████████████████████████████████████▏                | 15578/20000 [08:52<02:58, 24.81it/s]

 78%|███████████████████████████████████████████████████████████▏                | 15583/20000 [08:52<02:33, 28.71it/s]

 78%|███████████████████████████████████████████████████████████▏                | 15587/20000 [08:52<02:53, 25.39it/s]

 78%|███████████████████████████████████████████████████████████▏                | 15590/20000 [08:53<03:57, 18.56it/s]

 78%|███████████████████████████████████████████████████████████▎                | 15593/20000 [08:53<04:11, 17.55it/s]

 78%|███████████████████████████████████████████████████████████▎                | 15596/20000 [08:53<03:40, 19.99it/s]

 78%|███████████████████████████████████████████████████████████▎                | 15599/20000 [08:53<03:21, 21.86it/s]

 78%|███████████████████████████

 79%|████████████████████████████████████████████████████████████▏               | 15837/20000 [09:03<02:25, 28.64it/s]

 79%|████████████████████████████████████████████████████████████▏               | 15841/20000 [09:03<03:37, 19.13it/s]

 79%|████████████████████████████████████████████████████████████▏               | 15846/20000 [09:03<03:01, 22.92it/s]

 79%|████████████████████████████████████████████████████████████▏               | 15850/20000 [09:03<02:40, 25.92it/s]

 79%|████████████████████████████████████████████████████████████▏               | 15854/20000 [09:04<02:30, 27.52it/s]

 79%|████████████████████████████████████████████████████████████▎               | 15858/20000 [09:04<02:44, 25.11it/s]

 79%|████████████████████████████████████████████████████████████▎               | 15862/20000 [09:04<02:29, 27.72it/s]

 79%|████████████████████████████████████████████████████████████▎               | 15867/20000 [09:04<02:16, 30.24it/s]

 79%|███████████████████████████

 81%|█████████████████████████████████████████████████████████████▎              | 16119/20000 [09:13<02:05, 30.91it/s]

 81%|█████████████████████████████████████████████████████████████▎              | 16123/20000 [09:13<02:11, 29.56it/s]

 81%|█████████████████████████████████████████████████████████████▎              | 16127/20000 [09:14<02:14, 28.81it/s]

 81%|█████████████████████████████████████████████████████████████▎              | 16131/20000 [09:14<02:33, 25.14it/s]

 81%|█████████████████████████████████████████████████████████████▎              | 16134/20000 [09:14<02:32, 25.34it/s]

 81%|█████████████████████████████████████████████████████████████▎              | 16138/20000 [09:14<02:37, 24.54it/s]

 81%|█████████████████████████████████████████████████████████████▎              | 16141/20000 [09:14<02:31, 25.41it/s]

 81%|█████████████████████████████████████████████████████████████▎              | 16144/20000 [09:14<03:04, 20.89it/s]

 81%|███████████████████████████

 82%|██████████████████████████████████████████████████████████████▎             | 16392/20000 [09:23<02:08, 27.98it/s]

 82%|██████████████████████████████████████████████████████████████▎             | 16396/20000 [09:23<02:06, 28.58it/s]

 82%|██████████████████████████████████████████████████████████████▎             | 16401/20000 [09:23<01:51, 32.39it/s]

 82%|██████████████████████████████████████████████████████████████▎             | 16406/20000 [09:24<01:56, 30.89it/s]

 82%|██████████████████████████████████████████████████████████████▎             | 16411/20000 [09:24<01:50, 32.43it/s]

 82%|██████████████████████████████████████████████████████████████▍             | 16415/20000 [09:24<01:49, 32.83it/s]

 82%|██████████████████████████████████████████████████████████████▍             | 16419/20000 [09:24<01:51, 32.08it/s]

 82%|██████████████████████████████████████████████████████████████▍             | 16423/20000 [09:24<02:08, 27.76it/s]

 82%|███████████████████████████

 83%|███████████████████████████████████████████████████████████████▎            | 16658/20000 [09:33<02:19, 24.00it/s]

 83%|███████████████████████████████████████████████████████████████▎            | 16661/20000 [09:34<02:28, 22.47it/s]

 83%|███████████████████████████████████████████████████████████████▎            | 16665/20000 [09:34<02:08, 25.87it/s]

 83%|███████████████████████████████████████████████████████████████▎            | 16672/20000 [09:34<01:44, 31.85it/s]

 83%|███████████████████████████████████████████████████████████████▎            | 16677/20000 [09:34<01:55, 28.84it/s]

 83%|███████████████████████████████████████████████████████████████▍            | 16681/20000 [09:34<02:02, 27.07it/s]

 83%|███████████████████████████████████████████████████████████████▍            | 16687/20000 [09:34<01:46, 30.99it/s]

 83%|███████████████████████████████████████████████████████████████▍            | 16692/20000 [09:34<01:38, 33.56it/s]

 83%|███████████████████████████

 85%|████████████████████████████████████████████████████████████████▎           | 16926/20000 [09:44<01:58, 25.89it/s]

 85%|████████████████████████████████████████████████████████████████▎           | 16929/20000 [09:44<02:14, 22.91it/s]

 85%|████████████████████████████████████████████████████████████████▎           | 16932/20000 [09:44<02:07, 24.12it/s]

 85%|████████████████████████████████████████████████████████████████▎           | 16935/20000 [09:44<02:00, 25.43it/s]

 85%|████████████████████████████████████████████████████████████████▎           | 16938/20000 [09:44<01:56, 26.22it/s]

 85%|████████████████████████████████████████████████████████████████▍           | 16941/20000 [09:45<02:24, 21.22it/s]

 85%|████████████████████████████████████████████████████████████████▍           | 16945/20000 [09:45<02:03, 24.65it/s]

 85%|████████████████████████████████████████████████████████████████▍           | 16950/20000 [09:45<02:02, 24.85it/s]

 85%|███████████████████████████

 86%|█████████████████████████████████████████████████████████████████▎          | 17191/20000 [09:54<01:26, 32.60it/s]

 86%|█████████████████████████████████████████████████████████████████▎          | 17196/20000 [09:54<01:19, 35.31it/s]

 86%|█████████████████████████████████████████████████████████████████▎          | 17201/20000 [09:54<01:35, 29.16it/s]

 86%|█████████████████████████████████████████████████████████████████▍          | 17205/20000 [09:55<01:40, 27.83it/s]

 86%|█████████████████████████████████████████████████████████████████▍          | 17209/20000 [09:55<01:31, 30.40it/s]

 86%|█████████████████████████████████████████████████████████████████▍          | 17213/20000 [09:55<01:25, 32.76it/s]

 86%|█████████████████████████████████████████████████████████████████▍          | 17217/20000 [09:55<01:30, 30.65it/s]

 86%|█████████████████████████████████████████████████████████████████▍          | 17221/20000 [09:55<01:37, 28.44it/s]

 86%|███████████████████████████

 87%|██████████████████████████████████████████████████████████████████▎         | 17449/20000 [10:05<01:29, 28.62it/s]

 87%|██████████████████████████████████████████████████████████████████▎         | 17453/20000 [10:05<01:26, 29.49it/s]

 87%|██████████████████████████████████████████████████████████████████▎         | 17457/20000 [10:05<01:55, 22.09it/s]

 87%|██████████████████████████████████████████████████████████████████▎         | 17462/20000 [10:05<01:36, 26.27it/s]

 87%|██████████████████████████████████████████████████████████████████▎         | 17466/20000 [10:05<01:50, 22.87it/s]

 87%|██████████████████████████████████████████████████████████████████▍         | 17469/20000 [10:05<01:47, 23.46it/s]

 87%|██████████████████████████████████████████████████████████████████▍         | 17473/20000 [10:06<01:35, 26.35it/s]

 87%|██████████████████████████████████████████████████████████████████▍         | 17476/20000 [10:06<01:35, 26.33it/s]

 87%|███████████████████████████

 89%|███████████████████████████████████████████████████████████████████▍        | 17734/20000 [10:15<01:12, 31.29it/s]

 89%|███████████████████████████████████████████████████████████████████▍        | 17738/20000 [10:15<01:14, 30.48it/s]

 89%|███████████████████████████████████████████████████████████████████▍        | 17742/20000 [10:16<01:16, 29.53it/s]

 89%|███████████████████████████████████████████████████████████████████▍        | 17746/20000 [10:16<01:34, 23.75it/s]

 89%|███████████████████████████████████████████████████████████████████▍        | 17749/20000 [10:16<01:53, 19.79it/s]

 89%|███████████████████████████████████████████████████████████████████▍        | 17753/20000 [10:16<01:38, 22.80it/s]

 89%|███████████████████████████████████████████████████████████████████▍        | 17756/20000 [10:16<01:34, 23.84it/s]

 89%|███████████████████████████████████████████████████████████████████▍        | 17760/20000 [10:16<01:28, 25.36it/s]

 89%|███████████████████████████

 90%|████████████████████████████████████████████████████████████████████▎       | 17993/20000 [10:26<01:28, 22.72it/s]

 90%|████████████████████████████████████████████████████████████████████▍       | 17996/20000 [10:26<01:21, 24.50it/s]

 90%|████████████████████████████████████████████████████████████████████▍       | 17999/20000 [10:26<01:31, 21.83it/s]

 90%|████████████████████████████████████████████████████████████████████▍       | 18003/20000 [10:26<01:22, 24.22it/s]

 90%|████████████████████████████████████████████████████████████████████▍       | 18006/20000 [10:26<01:32, 21.54it/s]

 90%|████████████████████████████████████████████████████████████████████▍       | 18014/20000 [10:27<01:12, 27.56it/s]

 90%|████████████████████████████████████████████████████████████████████▍       | 18019/20000 [10:27<01:08, 28.77it/s]

 90%|████████████████████████████████████████████████████████████████████▍       | 18024/20000 [10:27<01:08, 28.95it/s]

 90%|███████████████████████████

 91%|█████████████████████████████████████████████████████████████████████▎      | 18244/20000 [10:36<01:06, 26.54it/s]

 91%|█████████████████████████████████████████████████████████████████████▎      | 18248/20000 [10:37<01:05, 26.89it/s]

 91%|█████████████████████████████████████████████████████████████████████▎      | 18252/20000 [10:37<00:59, 29.30it/s]

 91%|█████████████████████████████████████████████████████████████████████▎      | 18256/20000 [10:37<01:06, 26.27it/s]

 91%|█████████████████████████████████████████████████████████████████████▍      | 18260/20000 [10:37<01:03, 27.36it/s]

 91%|█████████████████████████████████████████████████████████████████████▍      | 18263/20000 [10:37<01:30, 19.22it/s]

 91%|█████████████████████████████████████████████████████████████████████▍      | 18267/20000 [10:37<01:18, 22.01it/s]

 91%|█████████████████████████████████████████████████████████████████████▍      | 18272/20000 [10:38<01:05, 26.28it/s]

 91%|███████████████████████████

 92%|██████████████████████████████████████████████████████████████████████▎     | 18496/20000 [10:47<01:05, 22.88it/s]

 92%|██████████████████████████████████████████████████████████████████████▎     | 18499/20000 [10:48<01:02, 24.15it/s]

 93%|██████████████████████████████████████████████████████████████████████▎     | 18504/20000 [10:48<00:59, 25.20it/s]

 93%|██████████████████████████████████████████████████████████████████████▎     | 18512/20000 [10:48<00:49, 29.93it/s]

 93%|██████████████████████████████████████████████████████████████████████▎     | 18516/20000 [10:48<00:55, 26.51it/s]

 93%|██████████████████████████████████████████████████████████████████████▍     | 18520/20000 [10:48<00:56, 25.97it/s]

 93%|██████████████████████████████████████████████████████████████████████▍     | 18523/20000 [10:48<00:56, 26.14it/s]

 93%|██████████████████████████████████████████████████████████████████████▍     | 18526/20000 [10:49<01:14, 19.87it/s]

 93%|███████████████████████████

 94%|███████████████████████████████████████████████████████████████████████▎    | 18760/20000 [10:58<00:51, 24.03it/s]

 94%|███████████████████████████████████████████████████████████████████████▎    | 18765/20000 [10:58<00:48, 25.60it/s]

 94%|███████████████████████████████████████████████████████████████████████▎    | 18769/20000 [10:58<00:45, 27.17it/s]

 94%|███████████████████████████████████████████████████████████████████████▎    | 18774/20000 [10:58<00:40, 30.20it/s]

 94%|███████████████████████████████████████████████████████████████████████▍    | 18783/20000 [10:59<00:32, 37.53it/s]

 94%|███████████████████████████████████████████████████████████████████████▍    | 18788/20000 [10:59<00:31, 38.74it/s]

 94%|███████████████████████████████████████████████████████████████████████▍    | 18793/20000 [10:59<00:49, 24.60it/s]

 94%|███████████████████████████████████████████████████████████████████████▍    | 18798/20000 [10:59<00:44, 26.94it/s]

 94%|███████████████████████████

 95%|████████████████████████████████████████████████████████████████████████▎   | 19031/20000 [11:09<00:45, 21.07it/s]

 95%|████████████████████████████████████████████████████████████████████████▎   | 19034/20000 [11:09<00:43, 22.12it/s]

 95%|████████████████████████████████████████████████████████████████████████▎   | 19037/20000 [11:09<00:42, 22.49it/s]

 95%|████████████████████████████████████████████████████████████████████████▎   | 19040/20000 [11:10<00:54, 17.46it/s]

 95%|████████████████████████████████████████████████████████████████████████▎   | 19044/20000 [11:10<00:45, 20.79it/s]

 95%|████████████████████████████████████████████████████████████████████████▍   | 19048/20000 [11:10<00:39, 24.27it/s]

 95%|████████████████████████████████████████████████████████████████████████▍   | 19054/20000 [11:10<00:32, 29.46it/s]

 95%|████████████████████████████████████████████████████████████████████████▍   | 19058/20000 [11:10<00:43, 21.43it/s]

 95%|███████████████████████████

 97%|█████████████████████████████████████████████████████████████████████████▎  | 19306/20000 [11:20<00:25, 27.19it/s]

 97%|█████████████████████████████████████████████████████████████████████████▍  | 19310/20000 [11:20<00:24, 27.69it/s]

 97%|█████████████████████████████████████████████████████████████████████████▍  | 19315/20000 [11:20<00:21, 31.43it/s]

 97%|█████████████████████████████████████████████████████████████████████████▍  | 19320/20000 [11:20<00:20, 33.58it/s]

 97%|█████████████████████████████████████████████████████████████████████████▍  | 19325/20000 [11:20<00:18, 36.99it/s]

 97%|█████████████████████████████████████████████████████████████████████████▍  | 19330/20000 [11:20<00:18, 35.42it/s]

 97%|█████████████████████████████████████████████████████████████████████████▍  | 19334/20000 [11:20<00:20, 32.94it/s]

 97%|█████████████████████████████████████████████████████████████████████████▍  | 19338/20000 [11:21<00:26, 24.91it/s]

 97%|███████████████████████████

 98%|██████████████████████████████████████████████████████████████████████████▍ | 19599/20000 [11:30<00:16, 23.94it/s]

 98%|██████████████████████████████████████████████████████████████████████████▍ | 19603/20000 [11:30<00:15, 25.74it/s]

 98%|██████████████████████████████████████████████████████████████████████████▌ | 19607/20000 [11:30<00:16, 23.86it/s]

 98%|██████████████████████████████████████████████████████████████████████████▌ | 19611/20000 [11:30<00:15, 24.48it/s]

 98%|██████████████████████████████████████████████████████████████████████████▌ | 19614/20000 [11:31<00:15, 25.31it/s]

 98%|██████████████████████████████████████████████████████████████████████████▌ | 19617/20000 [11:31<00:18, 20.21it/s]

 98%|██████████████████████████████████████████████████████████████████████████▌ | 19622/20000 [11:31<00:17, 22.17it/s]

 98%|██████████████████████████████████████████████████████████████████████████▌ | 19627/20000 [11:31<00:14, 26.49it/s]

 98%|███████████████████████████

 99%|███████████████████████████████████████████████████████████████████████████▍| 19855/20000 [11:41<00:06, 23.91it/s]

 99%|███████████████████████████████████████████████████████████████████████████▍| 19859/20000 [11:41<00:05, 25.56it/s]

 99%|███████████████████████████████████████████████████████████████████████████▍| 19863/20000 [11:41<00:05, 24.86it/s]

 99%|███████████████████████████████████████████████████████████████████████████▍| 19867/20000 [11:41<00:04, 28.03it/s]

 99%|███████████████████████████████████████████████████████████████████████████▌| 19871/20000 [11:42<00:06, 21.44it/s]

 99%|███████████████████████████████████████████████████████████████████████████▌| 19877/20000 [11:42<00:04, 25.50it/s]

 99%|███████████████████████████████████████████████████████████████████████████▌| 19881/20000 [11:42<00:04, 24.50it/s]

 99%|███████████████████████████████████████████████████████████████████████████▌| 19885/20000 [11:42<00:04, 25.29it/s]

 99%|███████████████████████████

In [30]:
def join_back(txt):
    new = ''
    for x in txt:
        new += ''.join(x)+" "
    new = new[:-1]
    return new

data[3] = data[3].progress_apply(lambda tokens: join_back(tokens))




  0%|                                                                                        | 0/20000 [00:00<?, ?it/s]


 13%|█████████▍                                                                | 2550/20000 [00:00<00:00, 25445.79it/s]


 29%|█████████████████████▏                                                    | 5713/20000 [00:00<00:00, 27001.43it/s]


 43%|███████████████████████████████▉                                          | 8632/20000 [00:00<00:00, 27619.07it/s]


 57%|█████████████████████████████████████████▎                               | 11321/20000 [00:00<00:00, 27393.09it/s]


 73%|█████████████████████████████████████████████████████▎                   | 14603/20000 [00:00<00:00, 28816.46it/s]


 87%|███████████████████████████████████████████████████████████████▌         | 17430/20000 [00:00<00:00, 28644.28it/s]


100%|█████████████████████████████████████████████████████████████████████████| 20000/20000 [00:00<00:00, 28825.63it/s]

In [32]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(data[3], data[6], test_size=0.25, random_state=42)

In [63]:
from sklearn.feature_extraction.text import CountVectorizer
count_vect = CountVectorizer()
X_train_counts = count_vect.fit_transform(X_train)
X_test_counts = count_vect.fit_transform(X_test)
print(X_train_counts.shape)

#TFIDF is an important feature as it prevents the longer documents to be more biased. 
from sklearn.feature_extraction.text import TfidfTransformer
tf_trans = TfidfTransformer()
X_train_tf = tf_trans.fit_transform(X_train_counts)
X_test_tf = tf_trans.fit_transform(X_test_counts)
print(X_train_tf.shape)

(15000, 37330)
(15000, 37330)


In [61]:
from sklearn.naive_bayes import MultinomialNB
clf = MultinomialNB().fit(X_train_tf, y_train)

In [70]:
#Defining a pipeline for quick operations in the later stages

from sklearn.pipeline import Pipeline
text_clf = Pipeline([('vect', CountVectorizer()),('tfidf', TfidfTransformer()),('clf', SGDClassifier(loss='hinge', penalty='l2',alpha=1e-3, n_iter=5, random_state=42))])

In [72]:
x = text_clf.fit(X_test, y_test)

D:\Anaconda3\lib\site-packages\sklearn\linear_model\stochastic_gradient.py:117: DeprecationWarning: n_iter parameter is deprecated in 0.19 and will be removed in 0.21. Use max_iter and tol instead.
  DeprecationWarning)


In [78]:
predicted = text_clf.predict(X_test)
print("Accuracy= ",np.mean(predicted == y_test)*100,"%")

Accuracy=  74.78 %


# Bigram performance:

## with ngram_range = (2,2) 
### Meaning "I love Data" -> [("I","love"), ("love","Data")]

In [88]:
test_vect = CountVectorizer(ngram_range=(2,2))
txt = "Sample text to check how the count vec with ngram range works"
test = test_vect.fit_transform([txt])
test_vect.get_feature_names()

#We can observe that the bigrams are working well

['check how',
 'count vec',
 'how the',
 'ngram range',
 'range works',
 'sample text',
 'text to',
 'the count',
 'to check',
 'vec with',
 'with ngram']

In [89]:
text_clf = Pipeline([('vect', CountVectorizer(ngram_range=(2,2))),('tfidf', TfidfTransformer()),('clf', SGDClassifier(loss='hinge', penalty='l2',alpha=1e-3, n_iter=5, random_state=42))])

In [90]:
x = text_clf.fit(X_test, y_test)

D:\Anaconda3\lib\site-packages\sklearn\linear_model\stochastic_gradient.py:117: DeprecationWarning: n_iter parameter is deprecated in 0.19 and will be removed in 0.21. Use max_iter and tol instead.
  DeprecationWarning)


In [91]:
predicted = text_clf.predict(X_test)
print("Accuracy= ",np.mean(predicted == y_test)*100,"%")

Accuracy=  50.56 %


## with ngram_range = (1,2) 
### Meaning "I love Data" -> [("I"), ("I","love"), ("love"),  ("love","Data"), ("Data")]

In [92]:
test_vect = CountVectorizer(ngram_range=(2,2))
txt = "Sample text to check how the count vec with ngram range works"
test = test_vect.fit_transform([txt])
test_vect.get_feature_names()

text_clf = Pipeline([('vect', CountVectorizer(ngram_range=(1,2))),('tfidf', TfidfTransformer()),('clf', SGDClassifier(loss='hinge', penalty='l2',alpha=1e-3, n_iter=5, random_state=42))])

x = text_clf.fit(X_test, y_test)

predicted = text_clf.predict(X_test)
print("Accuracy= ",np.mean(predicted == y_test)*100,"%")

D:\Anaconda3\lib\site-packages\sklearn\linear_model\stochastic_gradient.py:117: DeprecationWarning: n_iter parameter is deprecated in 0.19 and will be removed in 0.21. Use max_iter and tol instead.
  DeprecationWarning)


Accuracy=  76.84 %


# Conclusions
## Unigrams -> 74.78 %
## Bigrams ->  50.56 %
## Unigrams+Bigrams -> 76.84 %

## We can conclude that Unigrams+Bigrams seem to perform better in this Sample data.

# NOTE: This was done on a sampled data of 20,000 tuples with an equal amount of real and fake reviews. Actual performance on the full sized dataset may differ.